### PR TITLE
feat!: simplify `SignatureFunc` and add custom arg validation.

### LINKED ISSUE #675: refactor: simplify `SignatureFunc` from enum to struct
Now that we have `PolyFuncType` many opdefs can be declared using that. And of those that can't, many only use the binary function for validating type arguments, rather than for computing the signature. So instead we can have a simpler `SignatureFunc`:

first add a method to `CustomSignatureFunc` to return type params;
```rust
impl CustomSignatureFunc {
    fn compute_signature(...);

   fn params(&self) -> &[TypeParam];
}

```
```rust
struct SignatureFunc(Box<dyn CustomSignatureFunc> + 'static);

impl SignatureFunc {
    fn new_custom(impl CustomSignatureFunc + 'static) -> Self;
    fn new_typescheme(PolyFuncType) -> Self;   // compute_signature just calls instantiate
    fn new_typescheme_with_validation <T: &[Args] -> Result<(), SignatureError>>(PolyFuncType, T) -> Self; // compute_signature first calls validation closure then instantiate

}

```

### PR DESCRIPTION
**Pull request by ss2165**
Closes #675 but not exactly as specified in the issue - SignatureFunc is still an enum, but something close to the desired interface is achieved.

BREAKING_CHANGES: `Extension::add_op...` methods simplified and renamed, there is only `add_op` and `add_op_simple` now.

### CODE REVIEW

File: src/extension/op_def.rs [hunk 1]
```diff
@@ -175,19 +311,9 @@ impl OpDef {
...
-            SignatureFunc::CustomFunc {
-                static_params,
-                func,
-            } => {
-                let (static_args, other_args) = args.split_at(min(static_params.len(), args.len()));
-                check_type_args(static_args, static_params)?;
-                temp = func.compute_signature(&self.name, static_args, &self.misc, exts)?;
-                (&temp, other_args)
-            }
-        };
```
Comment (ss2165) [inline src/extension/op_def.rs:190]: moved to `SignatureFunc::compute_signature`

File: src/extension/prelude.rs [hunk 2]
```diff
@@ -8,13 +8,36 @@ use crate::{
...
 
-use super::ExtensionRegistry;
+use super::{ExtensionRegistry, SignatureError, SignatureFromArgs};
+struct ArrayOpCustom;
+
+const MAX: &[TypeParam; 1] = &[TypeParam::max_nat()];
+impl SignatureFromArgs for ArrayOpCustom {
+    fn compute_signature(&self, arg_values: &[TypeArg]) -> Result<PolyFuncType, SignatureError> {
+        let [TypeArg::BoundedNat { n }] = *arg_values else {
+            panic!("Should have been checked already.")
```
Comment (croyzor) [inline src/extension/prelude.rs:24]: This message could be more descriptive

File: src/extension/prelude.rs [hunk 3]
```diff
@@ -8,13 +8,36 @@ use crate::{
...
+        };
+        let elem_ty_var = Type::new_var_use(0, TypeBound::Any);
+
+        let var_arg_row = vec![elem_ty_var.clone(); n as usize];
+        let other_row = vec![array_type(elem_ty_var.clone(), TypeArg::BoundedNat { n })];
+
+        Ok(PolyFuncType::new(
+            vec![TypeParam::Type(TypeBound::Any)],
```
Comment (croyzor) [inline src/extension/prelude.rs:32]: Are we allowing arrays of linear types?

File: src/std_extensions/arithmetic/int_ops.rs [hunk 4]
```diff
@@ -11,33 +12,27 @@ use crate::{
...
-    if m > n {
-        return Err(SignatureError::InvalidTypeArgs);
-    }
-    Ok(FunctionType::new(
-        vec![int_type(arg0.clone())],
-        vec![int_type(arg1.clone())],
-    ))
+struct IOValidator {
+    // whether the first type argument should be greater than the second
+    f_gt_s: bool,
```
Comment (croyzor) [inline src/std_extensions/arithmetic/int_ops.rs:22]: amazing field name

File: src/std_extensions/logic.rs [hunk 5]
```diff
@@ -28,52 +27,53 @@ pub const OR_NAME: &str = "Or";
...
+fn logic_op_sig() -> impl SignatureFromArgs {
+    struct LogicOpCustom;
+
+    const MAX: &[TypeParam; 1] = &[TypeParam::max_nat()];
+    impl SignatureFromArgs for LogicOpCustom {
+        fn compute_signature(
+            &self,
+            arg_values: &[TypeArg],
+        ) -> Result<crate::types::PolyFuncType, crate::extension::SignatureError> {
+            let [TypeArg::BoundedNat { n }] = *arg_values else {
```
Comment (croyzor) [inline src/std_extensions/logic.rs:39]: Not sure what this line is for?

File: src/std_extensions/logic.rs [hunk 6]
```diff
@@ -28,52 +27,53 @@ pub const OR_NAME: &str = "Or";
 /// The extension identifier.
 pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("logic");
 
+fn logic_op_sig() -> impl SignatureFromArgs {
+    struct LogicOpCustom;
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/std_extensions/logic.rs:31]: not this is just vanilla rust crimes lol

File: src/extension/op_def.rs [hunk 7]
```diff
@@ -14,34 +14,80 @@ use crate::types::type_param::{check_type_args, TypeArg, TypeParam};
...
-    F: Fn(&[TypeArg]) -> Result<R, SignatureError> + Send + Sync,
-{
-    fn compute_signature(
-        &self,
-        _name: &SmolStr,
+/// Compute signature of `OpDef` given type arguments.
+pub trait SignatureFromArgs: Send + Sync {
+    /// Compute signature of node given
+    /// values for the type parameters.
+    fn compute_signature(&self, arg_values: &[TypeArg]) -> Result<PolyFuncType, SignatureError>;
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/extension/op_def.rs:37]: yeah in this function you have to bind all the "static" parameters. I will make that clearer in the comment.

File: src/extension/op_def.rs [hunk 8]
```diff
@@ -14,34 +14,80 @@ use crate::types::type_param::{check_type_args, TypeArg, TypeParam};
...
+
+/// Trait for validating type arguments to a PolyFuncType beyond conformation to
+/// declared type parameter, given just the arguments.
+pub trait ValidateJustArgs: Send + Sync {
+    /// Validate the type arguments of node given
+    /// values for the type parameters.
+    fn validate(&self, arg_values: &[TypeArg]) -> Result<(), SignatureError>;
+}
+
+impl<T: ValidateJustArgs> ValidateTypeArgs for T {
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/extension/op_def.rs:82]: That should be done all the time, this is just for _further_ validation - comment!

File: src/extension/op_def.rs [hunk 9]
```diff
@@ -440,18 +507,18 @@ mod test {
...
                     .collect();
                 Ok(PolyFuncType::new(
-                    vec![TP],
+                    vec![TP.to_owned()],
                     FunctionType::new(tvs.clone(), vec![Type::new_tuple(tvs)]),
                 ))
             }
+
+            fn static_params(&self) -> &[TypeParam] {
+                const MAX_NAT: &[TypeParam] = &[TypeParam::max_nat()];
```
[the remark this answers is not in the document]
Comment (ss2165) [reply src/extension/op_def.rs:516]: it is const, but I need a reference to a slice containing it. (I wouldn't be
able to do this if it _wasn't_ const already. )

[the remark this answers is not in the document]
Comment (croyzor) [reply src/extension/op_def.rs:516]: weird, i noticed the slice in other tests but not in this one

File: src/extension/prelude.rs (hunk 3 above)
Comment (ss2165) [reply src/extension/prelude.rs:32]: it was the case before - I haven't changed anything

File: src/std_extensions/logic.rs (hunk 5 above)
Comment (ss2165) [reply src/std_extensions/logic.rs:39]: added a comment

File: src/extension/prelude.rs (hunk 2 above)
Comment (ss2165) [reply src/extension/prelude.rs]: replaced with an eerror

File: src/extension/op_def.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (croyzor) [reply src/extension/op_def.rs:37]: If you have to bind all of the static parameters, then we should be able to substitute and return a `FunctionType`!

[the remark this answers is not in the document]
Comment (croyzor) [reply src/extension/op_def.rs:37]: Which can then `.into()` to a `PolyFuncType` if/when needed

[the remark this answers is not in the document]
Comment (ss2165) [reply src/extension/op_def.rs:37]: you are correct, I have undone this pointless refactor.

**Review by croyzor — APPROVED** [review]

**Pull request merged**